Ensure extra columns are not shown in table vis when showPartialRows:true

lukeelmers opened the pull request:
Closes elastic/kibana#22168 

## Summary

Resolves a regression where clicking `show partial rows` on the data table vis would automatically render columns with metrics at all levels -- even if you didn't check `metrics at all levels`.

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

~~- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)~~
~~- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)~~
~~- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~~
- [X] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
~~- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)~~

lukeelmers commented on src/ui/public/agg_response/tabify/tabify.js:
```diff
@@ -21,6 +21,17 @@ import _ from 'lodash';
+ * @param {boolean} respOpts.metricsAtAllLevels - reflects the value of vis.params.showMetricsAtAllLevels
+ * @param {boolean} respOpts.partialRows - reflects the value of vis.params.showPartialRows
+ * @param {Object} respOpts.timeRange - time range object, if provided
+ */
```
Since params are being thrown around willy-nilly and hard to track, I added this to reduce confusion for people working on Tabify in the future. (A TypeScript rewrite would be very helpful here).

lukeelmers commented on src/ui/public/vis/request_handlers/courier.js:
```diff
@@ -133,7 +134,8 @@ const CourierRequestHandlerProvider = function () {
       const tabifyParams = {
-        metricsAtAllLevels: isHierarchical,
+        isHierarchical,
+        metricsAtAllLevels,
```
These are split into separate params, because within table vis there is a distinction between the two. For example, the following is possible:
```javascript
vis.params.showMetricsAtAllLevels === false;
vis.params.showPartialRows === true;
vis.isHierarchical(); // still returns true
```

This is due to how we configure the table vis:
 https://github.com/elastic/kibana/blob/9b08fc719a9fb4c0c1dce55be73d395e7cd6b54a/src/core_plugins/table_vis/public/table_vis.js#L114-L116

ppisljar replied:
this is actually wrong i think .... 
`hierarchical === metricsAtAllLevels` (unless someone knows of any reasons it should be ? )

in courier request handler vis.isHierarchical() is used to figure out if we need to build a request with metrics on every level. (or it was, and its now refactored to look at metricsAtAllLevels parameter). we should use this same parameter in tabify (to figure out the same thing) ...
and partialRows should be completely independent ... as its just removing rows where we don't have value for every column.

elasticmachine commented:
Pinging @elastic/kibana-app

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/115/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/132/)

trevan commented:
I tested this and it looks good.

lukeelmers commented:
Thanks for reviewing @trevan! Much appreciated

ppisljar commented on src/ui/public/agg_response/tabify/_response_writer.js:
```diff
@@ -25,18 +25,27 @@ import { tabifyGetColumns } from './_get_columns';
  * @param {AggConfigs} aggs - the agg configs object to which the aggregation response correlates
- * @param {boolean} metricsAtAllLevels - setting to true will produce metrics for every bucket
- * @param {boolean} partialRows - setting to true will not remove rows with missing values
+ * @param {boolean} isHierarchical - reflects the value of vis.isHierarchical()
```
maybe we should also add description of what the flag does ?

ppisljar commented on src/ui/public/agg_response/tabify/_response_writer.js:
```diff
@@ -67,7 +76,7 @@ TabbedAggResponseWriter.prototype.row = function () {
   });
 
-  if (!toArray(this.rowBuffer).length || (!this.partialRows && this.isPartialRow(this.rowBuffer))) {
+  if (!toArray(this.rowBuffer).length || (!this.showPartialRows && this.isPartialRow(this.rowBuffer))) {
```
i would maybe revert this into `removePartialRows`

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2296/)

lukeelmers commented:
@ppisljar I renamed some of the params that are being passed around to reduce confusion.

Rather than using the `showMetricsAtAllLevels` vis param, which pretty much only exists for data table, I created a `columnsForAllBuckets` arg for tabify that specifically exists to indicate whether we need to have tabify return a column for each bucket/level.

I moved the logic for checking the existence of `params.showMetricsAtAllLevels` to the visualize data loader, which feels more appropriate than having tabify concern itself with the details of the vis.

The overall effect is the same, but hopefully this makes its purpose clearer so it doesn't get mixed up with `isHierarchical`.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2399/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2406/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2415/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2419/)

ppisljar commented:
hmm, maybe i am complicating too much, but to me the new naming does not actually work better ..
actually now we have 3 things .... metricsAtAllLevels, isHierarchical, clumnsForAllBcuekts .... and they all look super confusing to me :)

what about metricsAtAllLevels (which replaces isHierarchical and metricsAtAllLevels) ... and does just that, adds metric to every level in request (and then also knows how to use process this response in tabify)

partialRows: we decided not to go to the old way partial rows worked right ? we stay with definition that partial row is any row that is missing some column values.

so, minimalColumns/ however we call this/ just needs to remove all the metric columns on non-last levels. This also needs to happen after figuring out if row is partial or not. In the future, we don't want this setting in tabify .... there are other ways within canvas expressions to remove unneeded columns. (esaggs .... | columns exclude=1,2,3) or just by ignoring the rest of columns in the visualization (esaggs .... | tagcloud metric="5") <-- the fifth column is metric, ignore rest of the table

so again i would suggest to stash this for a while, i don't think this is a blocker bug.

trevan commented:
This is a reason why I can't upgrade to 6.5 unless I take @lukeelmers patch and fix it in my own copy.

timroes commented:
@trevan we missed 6.5.3 with that one, but it's still scheduled for a 6.5 patch release.

lukeelmers commented:
I agree we should still merge for a 6.5 patch release -- the work is done here; we just need to make sure the naming is making things less confusing rather than more confusing 😃 

And to @ppisljar's point, this will all become much simpler once #25996 lands.

> so, minimalColumns/ however we call this/ just needs to remove all the metric columns on non-last levels.

Correct.

> This also needs to happen after figuring out if row is partial or not.

Actually, it is currently possible (at least within table vis) to show partial rows AND columns for each bucket. Not sure why you would want to do this, but it is possible. So in that sense, I don't think determining partial rows is directly related to displaying columns for each bucket. You could have one or the other (or both) in the current implementation.

> partialRows: we decided not to go to the old way partial rows worked right ? we stay with definition that partial row is any row that is missing some column values.

Yes, that definition is how I've been thinking about `partialRows`, but I don't know how it used to work either...

> what about metricsAtAllLevels (which replaces isHierarchical and metricsAtAllLevels) ... and does just that, adds metric to every level in request (and then also knows how to use process this response in tabify)

In talking with @timroes just now, we realized that in the current tabify implementation, we aren't even using `metricsAtAllLevels` for anything other than checking whether to return columns for each bucket. So right now I think the only thing tabify actually *needs* to work is:
```
tabifyAggResponse(
  aggs,
  esResponse,
  {
    partialRows,
    timeRange,
    metricsAtAllLevels // <-- only used to determine which columns to return
  }
)
```

With that in mind, I would propose that rather than introducing new response options, we simply rename the `metricsAtAllLevels` argument (within tabify) to reflect its true purpose: whether we want to return columns for each bucket / metric or just for the last one.

We can keep the concept of isHierarchical / metricsAtAllLevels in other places like courier and visualize_data_loader. And we can take care of making sure the right value is passed to tabify from visualize_data_loader, which feels cleaner anyway as it will soon be replaced by pipeline_data_loader.

Then once we are switched over to canvas expressions we can simply remove that param from tabify since we know already it is just being used for one purpose.

@ppisljar LMK your thoughts and whether this approach makes sense to you. Then the question just becomes what to name this thing 😆 

`minimalColumns` ?
`columnsForAllBuckets` ?
`allColumns` ?
`giveMeAllOfTheThingz` ?
...something else?

lukeelmers commented:
I agree we should still merge for a 6.5 patch release -- the work is done here; we just need to make sure the naming is making things less confusing rather than more confusing 😃 

And to @ppisljar's point, this will all become much simpler once #25996 lands.

> so, minimalColumns/ however we call this/ just needs to remove all the metric columns on non-last levels.

Correct.

> This also needs to happen after figuring out if row is partial or not.

Actually, it is currently possible (at least within table vis) to show partial rows AND columns for each bucket. Not sure why you would want to do this, but it is possible. So in that sense, I don't think determining partial rows is directly related to displaying columns for each bucket. You could have one or the other (or both) in the current implementation.

> partialRows: we decided not to go to the old way partial rows worked right ? we stay with definition that partial row is any row that is missing some column values.

Yes, that definition is how I've been thinking about `partialRows`, but I don't know how it used to work either...

> what about metricsAtAllLevels (which replaces isHierarchical and metricsAtAllLevels) ... and does just that, adds metric to every level in request (and then also knows how to use process this response in tabify)

In talking with @timroes just now, we realized that in the current tabify implementation, we aren't even using `metricsAtAllLevels` for anything other than checking whether to return columns for each bucket. So right now I think the only thing tabify actually *needs* to work is:
```
tabifyAggResponse(
  aggs,
  esResponse,
  {
    partialRows,
    timeRange,
    metricsAtAllLevels // <-- only used to determine which columns to return
  }
)
```

With that in mind, I would propose that rather than introducing new response options, we simply rename the `metricsAtAllLevels` argument (within tabify) to reflect its true purpose: whether we want to return columns for each bucket / metric or just for the last one.

We can keep the concept of isHierarchical / metricsAtAllLevels in other places like courier and visualize_data_loader. And we can take care of making sure the right value is passed to tabify from visualize_data_loader, which feels cleaner anyway as it will soon be replaced by pipeline_data_loader.

Then once we are switched over to canvas expressions we can simply remove that param from tabify since we know already it is just being used for one purpose.

@ppisljar LMK your thoughts and whether this approach makes sense to you. Then the question just becomes what to name this thing 😆 

`minimalColumns` ?
`columnsForAllBuckets` ?
`allColumns` ?
`giveMeAllOfTheThingz` ?
...something else?

ppisljar commented:
to me personally, metricsAtEveryLevel tell exactly whats going on ....
minimalColumns .... whats minimal ?
columnsForAllBuckets .... buckets are columns as well .... ?
allColumns .... as oposed to 'not all columns` ? :)

however if i would need to vote, i would choose minimalColumns

so:
- isHierarchical property should be removed everywhere (has no use anymore)
- metricsAtAllLevels should signal to courier that we want metric at every level. the same flag needs to be passed to tabify as well (so it knows how to parse es response)
- partialRows: only passed to tabify, removes  rows which don't have values in every column
- minimalColumns: only passed to tabify, removes metrics on all levels but last (so does the reverse of metricsAtAllLevels)

still, it seems minimalColumns doesn't make much sense, as we are asking es to return metric at every level, just so we can remove it later .... why not just get a normal response (non metricsAtAllLevels ... or not hierarchical ...) ?
the only reason for this behaviour would be the partialRows (how it was implemented historically), howver it won't affect the partialRows implementation we have today (as we will be removing the columns before figuring out if the row is minimal or not) ... or am i missing something here ?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3640/)

lukeelmers commented:
retest

lukeelmers commented:
Pushed some updates:
- `partialRows` remains the same
- Added `minimalColumns`
- Removed `isHierarchical` from courier request handler, now using `metricsAtAllLevels` instead
- Tabify is not accepting `metricsAtAllLevels` as it currently does nothing with it... it only looks at `partialRows` (maybe this is something we should change eventually?) see here: https://github.com/elastic/kibana/blob/e5a94e7a1062f34471444ea29e20a844815842f6/src/ui/public/agg_response/tabify/tabify.js#L68-L71

There are some new issues with all of this behavior on `master` since #25996 was merged (console errors when trying to show partial rows in the UI). So I'm going to do a PR with these changes directly against `6.x` and backport to `6.5`. Then I will open a new PR for the fixes that need to happen on `master`, as that will require a slightly different fix.

@ppisljar Regarding your question about why we're getting metrics at every level for partial rows, here is how this was described when first introduced in #1926:

> Show partial rows
default: false
Allow rows that are incomplete to be shown. This causes the metrics to be calculated at every level because knowing what buckets exist without knowing anything about the contents seems pretty useless. This way, when a row is incomplete, we can show data from the last bucket we were able to create.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3661/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3677/)

lukeelmers commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3697/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3722/)

timroes commented on src/ui/public/agg_response/tabify/_response_writer.js:
```diff
@@ -25,22 +25,28 @@ import { tabifyGetColumns } from './_get_columns';
+} = {}) {
+  // Private
+  this._removePartialRows = !partialRows;
+  this._minimalColumns = minimalColumns;
```
Is there a specific reason we're storing that on a private variable? It seems to be only used locally in this function?

lukeelmers replied:
The goal here was to prevent confusion - I was surprised to find that some of the properties of the response writer were public (used in `tabify.js`) 

So the intent was to make it clear which items were part of the response writer’s contract, and which were just local implementation details and could be safely altered.

timroes commented on src/ui/public/visualize/loader/visualize_data_loader.ts:
```diff
@@ -73,10 +73,15 @@ export class VisualizeDataLoader {
     this.vis.showRequestError = false;
 
     try {
+      const minimalColumns =
```
Could we maybe add a short comment here, why we calculate `minimalColumns` like that, i.e. what are the use-cases we are trying to check for?

trevan commented:
@timroes, it looks good

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4241/)
